src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -132,28 +132,41 @@ Forward(const MatType& input, MatType& output)
 
   // Calculate the scores i.e. perform the matrix multiplication operation
   // on qProj and kProj. Here score = qProj . kProj'
-  scores = MultiplyCube2Cube(qProj, kProj, false, true);
+  scores = MultiplyCube2Cube(kProj, qProj, false, true);
 
   // Apply the attention mask if provided. The attention mask is used to black-
   // out future sequences and generally used in Encoder-Decoder attention.
   // The attention mask has elements -inf or 0.
   // The shape of the attention mask : (tgtSeqLen, srcSeqLen).
   if (!attnMask.is_empty())
   {
-    if (attnMask.n_rows != tgtSeqLen || attnMask.n_cols != srcSeqLen)
+    if (attnMask.n_slices != input.n_cols || attnMask.n_rows != tgtSeqLen || attnMask.n_cols != srcSeqLen)
       Log::Fatal << "The size of the 'attn_mask' is not correct.\n";
-    scores.each_slice() += attnMask;
+    for (size_t i = 0; i < batchSize; ++i)
+    {
+      for (size_t h = 0; h < numHeads; ++h)
+      {
+          scores.slice(i * numHeads + h) += attnMask.slice(i).t();
+      }
+    }
   }
 
+
   // Apply the key padding mask when provided. It blacks-out any particular
   // word in the sequence.
   // The key padding mask has elements -inf or 0
-  // The shape of keyPaddingMask : (1, srcSeqLen).
+  // The shape of keyPaddingMask : (batchSize, srcSeqLen).
   if (!keyPaddingMask.is_empty())
   {
-    if (keyPaddingMask.n_rows != 1 || keyPaddingMask.n_cols != srcSeqLen)
+    if (keyPaddingMask.n_rows != input.n_cols || keyPaddingMask.n_cols != srcSeqLen)
         Log::Fatal << "The size of the 'keyPaddingMask' is not correct.\n";
-    scores.each_slice() += repmat(keyPaddingMask, tgtSeqLen, 1);
+    for (size_t i = 0; i < batchSize; ++i)
+    {
+      for (size_t h = 0; h < numHeads; ++h)
+      {
+          scores.slice(i * numHeads + h) += repmat(keyPaddingMask.row(i), tgtSeqLen, 1).t();
```

Comment:
Same here, can we change `keyPaddingMask` to have shape `(srcSeqLen, batchSize)` instead?